core/src/main/java/org/eclipse/hono/connection/impl/ConnectionFactoryImpl.java

```diff
@@ -130,28 +131,53 @@ public void connect(
         final ProtonClient client = protonClient != null ? protonClient : ProtonClient.create(vertx);
         logger.debug("connecting to AMQP 1.0 container [{}://{}:{}]", clientOptions.isSsl() ? "amqps" : "amqp",
                 config.getHost(), config.getPort());
+
+        final AtomicBoolean connectionTimeoutReached = new AtomicBoolean(false);
+        final Long connectionTimeoutTimerId = config.getConnectTimeout() > 0
+                ? vertx.setTimer(config.getConnectTimeout(), id -> {
+                    if (connectionTimeoutReached.compareAndSet(false, true)) {
+                        failConnectionAttempt(clientOptions, connectionResultHandler, new TimeoutException(
```

Comment:
Using `java.util.concurrent.TimeoutException` here feels a little strange to me, in particular since its description states
> Exception thrown when a blocking operation times out. Blocking operations for which a timeout is specified need a means to indicate that the timeout has occurred.

In absence of another, better suited *standard* exception, we should IMHO define our own exception for this purpose ...

Reply:
Ok, I've used a newly defined `ConnectTimeoutException` now.